File: src/value/status.ts
export class GlobalStatus {
    public github: GitHubStatus;
    public projects: Map<string, ProjectStatus>;

    constructor() {
        this.projects = new Map<string, ProjectStatus>();
        this.github = new GitHubStatus();
    }
}

export class GitHubStatus {
    public token: string | null;
    public username: string | null;

    constructor() {
        this.token = null;
        this.username = null;
    }
}

export class CommandStatus {
    public cmd: string;
    public args: string[];

    public constructor(cmd: string) {
        this.cmd = cmd;
        this.args = [];
    }
}

export class ProjectStatus {
    public startupDependencies: string[];
    public name: string | null;
    public description: string | null;
    public buildCommand: CommandStatus | null;
    public runCommand: CommandStatus | null;
    public debugCommand: CommandStatus | null;
    public githubRepoUrl: string | null;
    public localRepoPath: string | null;
    public configuredGithubBranch: string | null;
    public backendProcessId: number | null;

    constructor() {
        this.name = null;
        this.githubRepoUrl = null;
        this.localRepoPath = null;
        this.configuredGithubBranch = null;
        this.buildCommand = null;
        this.debugCommand = null;
        this.runCommand = null;
        this.description = null;
        this.startupDependencies = [];
        this.backendProcessId = null;
    }
}

File: src-tauri/src/values/status.rs
use crate::common::{ProcessId, ProjectId};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GlobalStatus {
    pub github: GitHubStatus,
    pub projects: HashMap<ProjectId, ProjectStatus>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitHubStatus {
    pub username: Option<String>,
    pub token: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CommandStatus {
    pub cmd: String,
    pub args: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ProjectStatus {
    pub name: Option<String>,
    pub description: Option<String>,
    pub github_repo_url: Option<String>,
    pub github_branches: Vec<String>,
    pub configured_github_branch: Option<String>,
    pub local_repo_path: Option<PathBuf>,
    pub build_command: Option<CommandStatus>,
    pub run_command: Option<CommandStatus>,
    pub debug_command: Option<CommandStatus>,
    pub startup_dependencies: Vec<ProjectId>,
    pub backend_process_id: Option<ProcessId>,
}
